kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt

```diff
@@ -2453,33 +2415,21 @@ internal open class BufferedChannel<E>(
                 // This channel is already closed or cancelled; help to complete
                 // the closing or cancellation procedure.
                 completeCloseOrCancel()
-                // Clean the `prev` reference of the provided segment
-                // if all the previous cells are already covered by senders.
-                // It is important to clean the `prev` reference only in
-                // this case, as the closing/cancellation procedure may
-                // need correct value to traverse the linked list from right to left.
-                if (startFrom.id * SEGMENT_SIZE < sendersCounter) startFrom.cleanPrev()
                 // As the required segment is not found and cannot be allocated, return `null`.
                 null
             } else {
                 // Get the found segment.
                 val segment = it.segment
                 // Advance the `bufferEnd` segment if required.
                 if (!isRendezvousOrUnlimited && id <= bufferEndCounter / SEGMENT_SIZE) {
-                    bufferEndSegment.moveForward(segment)
+                    bufferEndSegment.moveToSpecifiedOrLast(id, bufferEndSegment.value)
```

Comment:
`moveToSpecifiedOrLast` is an inline function, which results in placing the whole implementation here. To avoid that, extracting `bufferEndSegment.moveToSpecifiedOrLast(id, bufferEndSegment.value)` to a separate method would help.